### PR TITLE
MueLu: regionMG use hierarchy and level containers

### PR DESCRIPTION
**Pull request by pohm01**
@trilinos/muelu 

## Motivation

Update the regionMG code to use Hierarchy and level containers to pass information, reducing the number of input parameters for the region functions.

## Testing
The region tests pass with same output as before these changes.

### CODE REVIEW

File: packages/muelu/research/regionMG/src/SetupRegionHierarchy_def.hpp [hunk 1]
```diff
@@ -1099,22 +1045,24 @@ void vCycle(const int l, ///< ID of current level
       bool coarseZeroInitGuess = true;
 
       // Call V-cycle recursively
-      vCycle(l+1, numLevels, cycleType,
-             coarseRegX, coarseRegB, regMatrices, regProlong, compRowMaps,
-             quasiRegRowMaps, regRowMaps, regRowImporters, regInterfaceScalings,
+      vCycle(l+1, numLevels, cycleType, regHierarchy,
+             coarseRegX, coarseRegB,// regMatrices, regProlong, compRowMaps,
+             //quasiRegRowMaps, regRowMaps, regRowImporters, regInterfaceScalings,
              smootherParams, coarseZeroInitGuess, coarseSolverData, hierarchyData);
```
Comment (mayrmt) [inline packages/muelu/research/regionMG/src/SetupRegionHierarchy_def.hpp:1051]:
```suggestion
      vCycle(l+1, numLevels, cycleType, regHierarchy,
             coarseRegX, coarseRegB,
             smootherParams, coarseZeroInitGuess, coarseSolverData, hierarchyData);
```

File: packages/muelu/research/regionMG/test/structured/Driver_Structured_Regions.cpp [hunk 2]
```diff
@@ -687,58 +687,82 @@ int main_(Teuchos::CommandLineProcessor &clp, Xpetra::UnderlyingLib& lib, int ar
...
-  const int numLevels = regProlong.size();
+  //const int numLevels = regProlong.size();
+  const int numLevels = regHierarchy->GetNumLevels();
 
   // Set data for fast MatVec
   // for (auto levelSmootherParams : smootherParams) {
   for(LO levelIdx = 0; levelIdx < numLevels; ++levelIdx) {
+    RCP<MueLu::Level> level = regHierarchy->GetLevel(levelIdx);
+    level->print( std::cout, MueLu::Extreme );
+    std::cout<<"ababababababababababababa"<<std::endl;
```
Comment (mayrmt) [inline packages/muelu/research/regionMG/test/structured/Driver_Structured_Regions.cpp:759]: Remove debug output?

File: packages/muelu/research/regionMG/test/structured/Driver_Structured_Regions.cpp [hunk 3]
```diff
@@ -849,13 +873,17 @@ int main_(Teuchos::CommandLineProcessor &clp, Xpetra::UnderlyingLib& lib, int ar
...
+    RCP<MueLu::Level> level = regHierarchy->GetLevel(0);
+    RCP<Xpetra::Vector<Scalar, LocalOrdinal, GlobalOrdinal> > regInterfaceScalings1 = level->Get<RCP<Xpetra::Vector<Scalar, LocalOrdinal, GlobalOrdinal> > >("regInterfaceScalings");
       // check for convergence
       {
         ////////////////////////////////////////////////////////////////////////
         // SWITCH BACK TO NON-LEVEL VARIABLES
         ////////////////////////////////////////////////////////////////////////
         computeResidual(regRes, regX, regB, regionMats, *smootherParams[0]);
-        scaleInterfaceDOFs(regRes, regInterfaceScalings[0], true);
+        //scaleInterfaceDOFs(regRes, regInterfaceScalings[0], true);
```
Comment (mayrmt) [inline packages/muelu/research/regionMG/test/structured/Driver_Structured_Regions.cpp:885]: Remove?

File: packages/muelu/research/regionMG/test/structured/Driver_Structured_Regions.cpp [hunk 4]
```diff
@@ -880,11 +908,12 @@ int main_(Teuchos::CommandLineProcessor &clp, Xpetra::UnderlyingLib& lib, int ar
...
-        vCycle(0, numLevels, cycleType,
-               regCorrect, regRes, regMatrices,
-               regProlong, compRowMaps, quasiRegRowMaps, regRowMaps, regRowImporters,
-               regInterfaceScalings, smootherParams, zeroInitGuess, coarseSolverData, hierarchyData);
+        scaleInterfaceDOFs(regRes, regInterfaceScalings1, false);
+        vCycle(0, numLevels, cycleType, regHierarchy,
+               regCorrect, regRes,// regMatrices,
+               //regProlong, compRowMaps, quasiRegRowMaps, regRowMaps, regRowImporters,
+               //regInterfaceScalings,
+               smootherParams, zeroInitGuess, coarseSolverData, hierarchyData);
```
Comment (mayrmt) [inline packages/muelu/research/regionMG/test/structured/Driver_Structured_Regions.cpp:916]:
```suggestion
        vCycle(0, numLevels, cycleType, regHierarchy,
               regCorrect, regRes,
               smootherParams, zeroInitGuess, coarseSolverData, hierarchyData);
```

**Review by mayrmt — CHANGES_REQUESTED** [review]
@pohm01  Thanks for your work! Changes look good to me. 

**Requested changes:** I highlighted some spots, where there seems to be debug output to be removed before merge. Please also remove (and not just comment) outdated code to keep this already quite messy part of the code as clean as possible.

**Further recommendations:** You could also make more extensive use of MueLu's/Xpetra's short names predefined in `MueLu_UseShortNames.hpp`, such that you do not have to state namespace and all template parameters for every `Vector`, `Matrix`, `Hierarchy`, ... object. In many places, this probably would make the code more readable.

**Tests:** I ran all region MG tests on my machine. They take a long time (cf. #8374), but they pass.

File: packages/muelu/research/regionMG/src/SetupRegionHierarchy_def.hpp [hunk 5]
```diff
@@ -133,40 +133,39 @@ void createContinuousCoarseLevelMaps(const RCP<const Xpetra::Map<LocalOrdinal, G
...
-                         Array<Teuchos::RCP<Xpetra::MultiVector<LocalOrdinal, LocalOrdinal, GlobalOrdinal, Node> > >& regionsPerGIDWithGhosts,
-                         Array<Teuchos::ArrayRCP<LocalOrdinal> >& regionMatVecLIDs,
-                         Array<Teuchos::RCP<Xpetra::Import<LocalOrdinal, GlobalOrdinal, Node> > >& regionInterfaceImporter) {
+                         Teuchos::RCP<MueLu::Hierarchy<Scalar, LocalOrdinal, GlobalOrdinal, Node> > regHierarchy) {
 
 #include "Xpetra_UseShortNames.hpp"
 
   using MT = typename Teuchos::ScalarTraits<SC>::magnitudeType;
 
+  RCP<MueLu::Level> level0 = regHierarchy->GetLevel(0);
```
[the remark this answers is not in the document]
Comment (pohm01) [reply packages/muelu/research/regionMG/src/SetupRegionHierarchy_def.hpp:142]: I need the MueLu namespace here. Is it better to do it this way or to include `MueLu_UseShortNames.hpp` in each function?

**Review by lucbv — APPROVED** [review]
This looks fine, I think the use of the `createPreconditionerXpetra` to generate the initial MueLu hierarchy before setting it up later.
As discussed we can have a follow on PR which removes maps from the function interfaces as they are most likely already included in the `Import` and `Matrix/Vector` objects.

**Review by mayrmt — APPROVED** [review]
Recent changes look good to me.

**Review by mayrmt — APPROVED** [review]

**Pull request merged**